Feature/speedy ghcn

ClaraDraper-NOAA opened the pull request:
## Description

This PR speeds up the ghcn ioda converter. The original version takes about 90 seconds, and this version takes about 1 second.

It also includes some tidy up of the code, and removes the maskout option. 

Additionally, the previous code would exclude data from stations not included in the fix file (the fix files provides the location of the station). In the previous test case, this means just 78 of 2000 stations are saved. Missing stations from the fix file usually means that the wrong fix file has been used, so this PR also adds code to exit if there are missing stations. There is an optional argument that can be used to over-ride this, and simply warn on missing stations.

## Issue(s) addressed

Addresses issue [1677](https://github.com/JCSDA-internal/ioda-converters/issues/1677)
## Dependencies

None.

## Impact

None

## Checklist

- [ x] I have performed a self-review of my own code
- [ x] I have made corresponding changes to the documentation
- [x ] I have run the unit tests before creating the PR

NOTE: see comment on fix files above. The old test had a serious bug, so I have updated it. If I change the code in this PR to exclude stations missing from the fix file I can reproduce the old test.

ClaraDraper-NOAA committed:
1) removed assignValue function to search for station ID in data file, and replaced with panda's merge.
2) removed the maskout option, and coded to always only include stations with data
3) clean-up: removed redundant intializations

-change 1 speeds up the converter by a factor of ~100.

ClaraDraper-NOAA committed:
Changes so that test case runs.

Note: test is failing, but I think the new results are correct.

ClaraDraper-NOAA commented:
Creating a draft for now as the test is failing. The original code generates an IODA file with only 78 obs, while the updated code has 2000. I think the updated code is correct, but I am still looking into it.

ClaraDraper-NOAA commented:
FYI: @BenjaminRuston . I've left this as a draft for now, but will convert it once I figure out what is going on with the test.

CoryMartin-NOAA reviewed (commented):
do you need to update the test reference, @ClaraDraper-NOAA, now that maskout is removed?

ClaraDraper-NOAA commented:
> do you need to update the test reference, @ClaraDraper-NOAA, now that maskout is removed?

No. The test used maskout=True, which I've now hardcoded as the only option. This update should have passed the existing test ( I think I know why it's not though).

ClaraDraper-NOAA committed:
Update code to exit (default) or warn if there are stations in the data
file that are not in the station_list. This usually means that the station
list needs to be updated.

ClaraDraper-NOAA committed:
new input files for test case, with consistent station list and data files.

ClaraDraper-NOAA committed:
Added new test output.

ClaraDraper-NOAA commented:
The test was failing due to a bug in the old test. Specifically, the old test excludes data from stations not included in the fix file (the fix files provides the location of the station). In the old test case, this means just 78 of 2000 stations are saved. Missing stations from the fix file usually means that the wrong fix file has been. 

I have replaced the fix file and input file to be from the latest download of the data. Now, of the 2000 stations in the input file 1999 stations are saved (one is removed due to invalid snow depth). I've also added a check that will exit if there are stations missing from the fix files, since we do expect them all to be there.

ClaraDraper-NOAA commented:
@CoryMartin-NOAA @yuanxue2870 @YoulongXia-NOAA @BenjaminRuston 

This is ready to review (and is needed for the reanalysis prototype runs at PSL).

ClaraDraper-NOAA committed:
Updated readme.

yuanxue2870 reviewed (approved):
Looks good to me. Thanks!

YoulongXia-NOAA reviewed (approved):
look good. I approved this PR.

BenjaminRuston committed:
Merge branch 'develop' into feature/speedy_ghcn

BenjaminRuston commented:
@ClaraDraper-NOAA my bad... the branch need to be put into JCSDA-internal!

I can do this for you,, but our CI can't checkout code from NOAA github:
`ClaraDraper-NOAA:feature/speedy_ghcn`

BenjaminRuston commented:
close as duplicate of https://github.com/JCSDA-internal/ioda-converters/pull/1681
